src/bootstrap.rs

```diff
@@ -30,3 +30,60 @@ pub fn start() {
         let _ = connect::connect_to(proxy, None, Some(&maker));
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::test_utils::new_random_qp2p_for_unit_test;
+    use crate::{Builder, Config, Event};
+    use env_logger;
+    use std::collections::HashSet;
+    use std::net::{IpAddr, Ipv4Addr};
+    use std::sync::mpsc;
+
+    // Try to bootstrap to 4 different peer at once. Use an artificial delay for 3 of the peers.
+    // Make sure in the end we have only one bootstrapped connection (all other connections are
+    // supposed to be dropped).
+    #[test]
+    fn bootstrap_to_only_one_node() {
+        env_logger::init();
+
+        let mut bs_nodes = Vec::new();
+        let mut bs_peer_addrs = Vec::new();
+        let mut hcc_contacts = HashSet::new();
+
+        // Spin up 4 peers that we'll use for hardcoded contacts.
+        for i in 0..4 {
+            let (tx, rx) = mpsc::channel();
+            let mut qp2p = unwrap!(Builder::new(tx)
+                .with_config(Config {
+                    ip: Some(IpAddr::V4(Ipv4Addr::LOCALHOST)),
+                    port: Some(0),
+                    ..Default::default()
+                })
+                .build());
+
+            qp2p.set_connect_delay(i * 30);
+
+            let ci_info = unwrap!(qp2p.our_connection_info());
+
+            bs_peer_addrs.push(ci_info.peer_addr);
+            hcc_contacts.insert(ci_info);
+            bs_nodes.push((rx, qp2p));
+        }
+
+        // Create a client and bootstrap to 4 of the peers we created previously.
+        let (mut qp2p0, rx0) = new_random_qp2p_for_unit_test(false, hcc_contacts);
+        qp2p0.bootstrap();
+
+        match unwrap!(rx0.recv()) {
+            Event::BootstrappedTo { node, .. } => {
+                assert_eq!(node.peer_addr, bs_peer_addrs[0]);
+            }
+            ev => panic!("Unexpected event: {:?}", ev),
+        }
+
+        // We expect only 1 bootstrap connection to succeed.
```

Comment:
Well you have put a delay for the other connections, so even if they matured and there was a bug in production to not eject them, the test would have ended by then. So i'm not getting what exactly is being tested.